### DISCUSSION TITLE
Get operations from database for Fortran interface

### DISCUSSION THREAD

**Discussion by pcazade** [spglib/spglib#529] [General]
Dear Developers,

Would it be possible to add a function to the Fortran interface to get the rotations and translations operations from the database using the Hall number? I implemented space group symmetry in CP2K using Spglib a few years ago and am working on improving its performances using the dataset. Still, the standard procedure has too much overhead for very large systems (I am modelling protein crystals with more than 1 million atoms). Being able to directly access the symmetry operations with the Hall number would reduce this overhead.

Thank you in advance for your help.

Best regards,
Pierre

**Comment by atztogo** [comment]
Do you mean `spg_get_symmetry_from_database`?
https://spglib.readthedocs.io/en/latest/api.html#spg-get-symmetry-from-database

> Being able to directly access the symmetry operations with the Hall number would reduce this overhead.

Why do you think so?

**Reply by LecrisUT** [reply]
I think the idea is good, e.g. if you know beforehand what lattice structure you have and you only want to get the symmetry operations. I am not sure about the interface though, I think it would be more valuable to expand it to be able to query via any parameter like the international symbol and such.

**Reply by pcazade** [reply]
Yes. I mean something like spg_get_symmetry_from_database. It does not seem to be provided by the Fortran interface. I use this function in some of my Python scripts and was surprised not to find it available for Fortran.

**Reply by atztogo** [reply]
> I think the idea is good, e.g. if you know beforehand what lattice structure you have

This was also my question. It means it is almost equivalent to that you know symmetry operations already beforehand. So I have no idea about the use case.

> was surprised not to find it available for Fortran.

Functions in the Fortran interface have been written on demand. You can make it and give us PR if you want. Otherwise it is me who should work on it.

**Reply by LecrisUT** [reply]
> This was also my question. It means it is almost equivalent to that you know symmetry operations already beforehand. So I have no idea about the use case.

Theoretically, yes they know it, but computationally they don't have access to it and we shouldn't encourage them to reinvent the wheel when all of this is already accessible in `spglib`.

**Reply by pcazade** [reply]
>This was also my question. It means it is almost equivalent to that you know symmetry operations already beforehand. So I have no idea about the use case.

Getting access to the symmetry operations to enforce them during the optimization process. I don't see an issue with either you don't know the space group and you let spglib find it, or you do and speed up the process by directly accessing the symmetry operations. Without such a function (which exists in Python), I would have to hard code in CP2K the list of symmetry operations for the 530 Hall symmetry group which is rather unnecessary.

**Reply by LecrisUT** [reply]
@pcazade are you up to writing a PR for this, you can reference [spg_get_dataset](https://github.com/spglib/spglib/blob/d9c572d1986b0eccc4ed5f18c8d827729a320313/fortran/spglib_f08.F90#L767-L774) for the design and make sure the column/row majorness is respected when you move from C to Fortran. Otherwise, I don't think I can come back to this one within this (coming) month due to hectic schedule. I can review it though.

**Reply by pcazade** [reply]
@LecrisUT I will work on it and send it to you for review.

**Reply by pcazade** [reply]
The wrapper is ready and tested. I did a git commit and I am trying to push it to the GitHub just to be denied access. I used gh repo clone and did all the account setup it asks for, so I am confused as to why I cannot push. I have not used Github in four years so I am not up to speed with the recent requirements.

**Reply by LecrisUT** [reply]
First fork `spglib`
![Screenshot_20241002_135308](https://github.com/user-attachments/assets/33b1e3e9-32ba-46a7-be4e-102d8a8d1fc4)

Next make sure your commits are in a separate branch, if you were writing it on `develop`, than you can just run:
```console
$ git checkout -b whatever_name_you_want
```
from the branch where you wrote it.

Then you add your fork as a remote and push to there:
```console
If you have an ssh key setup, use the one below, otherwise use the second one and you should be prompted your password or token
$ git remote add pcazade git@github.com:pcazade/spglib
$ git remote add pcazade https://github.com/pcazade/spglib
$ git fetch pcazade
$ git push pcazade
```

**Comment by pcazade** [comment]
Hi Cristian,

Thanks for your help. The changes have been pushed into a new branch
(database_for_fortran) in the fork i created. Should I just know press
"Compare & pull request" on Github?

Regards,
Pierre

**Reply by LecrisUT** [reply]
Yes, I don't have on the top of my head what the text should read like, but I believe that's the one.
